src/registry/registry.c

```diff
@@ -0,0 +1,164 @@
+//
+// registry.c
+//
+// Copyright (c) 2020 clib authors
+// MIT licensed
+//
+
+#include "github-registry.h"
+#include "gitlab-registry.h"
+#include "gumbo-parser/gumbo.h"
+#include "list/list.h"
+#include "registry-internal.h"
+#include "url/url.h"
+#include <stdlib.h>
+#include <string.h>
+
+enum registry_type_t {
+  REGISTRY_TYPE_GITHUB,
+  REGISTRY_TYPE_GITLAB,
+};
+
+struct registry_t {
+  enum registry_type_t type;
+  char *url;
+  char *hostname;
+  char *secret;
+  list_t *packages;
+};
+
+/**
+ * Create a new registry package.
+ */
+registry_package_ptr_t registry_package_new() {
+  registry_package_ptr_t pkg = malloc(sizeof(struct registry_package_t));
+  if (pkg) {
+    pkg->id = NULL;
+    pkg->href = NULL;
+    pkg->description = NULL;
+    pkg->category = NULL;
+  }
+  return pkg;
+}
+
+/**
+ * Release the memory held by the package.
+ */
+void registry_package_free(registry_package_ptr_t pkg) {
+  free(pkg->id);
+  free(pkg->href);
+  free(pkg->description);
+  free(pkg->category);
+  free(pkg);
+}
+
+registry_ptr_t registry_create(const char *url, const char *secret) {
+  registry_ptr_t registry = malloc(sizeof(struct registry_t));
+  registry->url = strdup(url);
+  registry->secret = strdup(secret);
+
+  if (strstr(url, "github.com") != NULL) {
+    registry->type = REGISTRY_TYPE_GITHUB;
+  } else if (strstr(url, "gitlab") != NULL) {
+    registry->type = REGISTRY_TYPE_GITLAB;
+  } else {
+    registry_free(registry);
```

Comment:
Maybe add a warning message that something is wrong?

Reply:
I have added a warning, what do you think?

Reply:
Awesome!